fdbbackup/FileDecoder.actor.cpp

```diff
@@ -217,54 +230,74 @@ struct DecodeProgress {
 
 	// The following are private APIs:
 
+	// Returns true if value contains complete data.
+	bool isValueComplete(StringRef value) {
+		StringRefReader reader(value, restore_corrupted_data());
+
+		reader.consume<uint64_t>(); // Consume the includeVersion
+		uint32_t val_length = reader.consume<uint32_t>();
+		return val_length == value.size() - sizeof(uint64_t) - sizeof(uint32_t);
+	}
+
 	// PRECONDITION: finished() must return false before calling this function.
 	// Returns the next batch of mutations along with the arena backing it.
 	ACTOR static Future<VersionedMutations> getNextBatchImpl(DecodeProgress* self) {
 		ASSERT(!self->finished());
 
-		state std::pair<Arena, KeyValueRef> arena_kv = self->keyValues[0];
-
-		// decode this batch's version
-		state std::pair<Version, int32_t> version_part = decode_key(arena_kv.second.key);
-		ASSERT(version_part.second == 0); // first part number must be 0.
-
-		// decode next versions, check if they are continuous parts
-		state int idx = 1; // next kv pair in "keyValues"
-		state int bufSize = arena_kv.second.value.size();
-		state int lastPart = 0;
 		loop {
-			// Try to decode another block if needed
-			if (idx == self->keyValues.size()) {
-				wait(readAndDecodeFile(self));
+			state std::tuple<Arena, Version, int32_t, StringRef> tuple = self->keyValues[0];
+
+			ASSERT(std::get<2>(tuple) == 0); // first part number must be 0.
+
+			// decode next versions, check if they are continuous parts
+			state int idx = 1; // next kv pair in "keyValues"
+			state int bufSize = std::get<3>(tuple).size();
+			state int lastPart = 0;
+			loop {
+				// Try to decode another block if needed
+				if (idx == self->keyValues.size()) {
+					wait(readAndDecodeFile(self));
+				}
+				if (idx == self->keyValues.size()) break;
```

Comment:
I know this is a draft but this line combined with the `if` on line 258 seems like a problem.

Reply:
This intentional. At line 258's if block, we try to decode more blocks from input file. here if there is no more data, we break.

Reply:
why not combining the two blocks?

Reply:
I don't see an easy way to combine these two blocks without an extra variable. I think current logic is cleaner and readable.

Reply:
```
if (idx == self->keyValues.size()) {
 	wait(readAndDecodeFile(self));
        break;
}
```

Isn't this the same as the old code?

Reply:
It's not the same. `readAndDecodeFile` modifies `self->keyValues`, so the second check is needed.

Reply:
I see. I didn’t realize that side effect.